File: src/com/easy/MissingNumber_268.java
/**
 * 
 */
package com.easy;

/**
 * @author Chris Lin
 * @version 1.1
 */
public class MissingNumber_268 {

    // simple calculation
    // Time Complexity: O(n)
    // Space Complexity:O(1)
    // Runtime: 0 ms, faster than 100.00%
    // Memory Usage: 41 MB, less than 61.27%
    public int missingNumber(int[] nums) {
        int mResult = 0;
        mResult = (nums.length + 1) * nums.length / 2;

        for (int num : nums) {
            mResult -= num;
        }

        return mResult;
    }

    // XOR operation
    // Time Complexity: O(n)
    // Space Complexity:O(1)
    // Runtime: 0 ms, faster than 100.00%
    // Memory Usage: 41.3 MB, less than 59.68%
    public int missingNumber2(int[] nums) {
        int mResult = 0;
        for (int i = 0; i < nums.length; i++) {
            mResult = mResult ^ i ^ nums[i];
        }
        // need XOR array biggest number ex: [0,1] so
        // length is 2
        mResult ^= nums.length;

        return mResult;
    }
}

File: Cpp_src/easy/missing_number_268.cc
#include <algorithm>
#include <iostream>
#include <vector>
using namespace std;

// Use bit manipulation
// Time Complexity: O(n)
// Space Complexity:O(1)
// Runtime: 32 ms, faster than 21.72%
// Memory Usage: 9.7 MB, less than 98.04%
class Solution
{
public:
    int missingNumber(vector<int> &nums)
    {
        int result = nums.size();
        for (int i = 0; i < nums.size(); i++)
        {
            result ^= i;
            result ^= nums[i];
        }
        return result;
    }
};

// Use sum
// Time Complexity: O(n)
// Space Complexity:O(1)
// Runtime: 24 ms, faster than 82.06%
// Memory Usage: 9.9 MB, less than 66.67%
class Solution2
{
public:
    int missingNumber(vector<int> &nums)
    {
        int sum = nums.size() * (nums.size() + 1) / 2;
        for (int num : nums)
        {
            sum -= num;
        }
        return sum;
    }
};

// Use binary search
// Time Complexity: O(nlogn)
// Space Complexity:O(1)
// Runtime: 32 ms, faster than 21.72%
// Memory Usage: 9.9 MB, less than 72.55%
class Solution3
{
public:
    int missingNumber(vector<int> &nums)
    {
        sort(nums.begin(), nums.end());
        int index_left = 0, index_right = nums.size();
        while (index_left < index_right)
        {
            int index_mid = index_left + (index_right - index_left) / 2;
            if (nums[index_mid] > index_mid)
            {
                index_right = index_mid;
            }
            else
            {
                index_left = index_mid + 1;
            }
        }
        return index_left;
    }
};